onnxruntime/core/framework/provider_bridge_ort.cc

```diff
@@ -504,12 +504,26 @@ struct ProviderHostImpl : ProviderHost {
 
 struct ProviderLibrary {
   ProviderLibrary(const char* filename) {
-
     std::string full_path = Env::Default().GetRuntimePath() + std::string(filename);
     Env::Default().LoadDynamicLibrary(full_path, &handle_);
     if (!handle_)
       return;
 
+#ifdef _WIN32
+    {
+      // We crash when unloading DNNL on windows due to OpenMP (As there are threads running code inside the openmp runtime DLL if
+      // OMP_WAIT_POLICY is set to ACTIVE). To avoid this, we pin the OpenMP DLL so that it unloads as late as possible.
+      HMODULE handle{};
```

Comment:
Do we know what the threads are running that depends on or attempts to touch something in the EP DLL? I'm wondering if there's some cleanup missing on our (or in the DNNL code) side that could also fix this problem. Doesn't quite make sense that this sort of pinning approach would be needed in what seems like a fairly standard usage of openmp.

Reply:
The openmp threads that created by the runtime.

Reply:
The threads aren't touching EP code, they're running OpenMP code. When the DNNL provider unloads, it's dependencies also unload (which includes the OpenMP library) which now leaves threads running unloaded code and crashes.

@snnn , do you know if OpenMP supports unloading DLLs that use it at runtime or only at process exit time?

Reply:
For the answer I got from VC++ team, it's not supported. Do not use FreeLibrary to unload the the DLL if the wait policy is active. If it is passive, it is ok, but please use it as it is, there is no guarantee.